### ISSUE TITLE
Auto-normalization and CLIP embeddings

### ISSUE THREAD

**Issue by marunine** [lucidrains/DALLE2-pytorch#102]
I have an x-clip model trained on normalized image data in the range `[-1, 1]` with good loss performance on a custom dataset.

I've tried to re-use the same custom dataset/dataloader that provides normalized images with the models in this repo but have run into issues with the auto-normalization.

If I leave the images in the range `[-1, 1]` then the `forward` pass in the Decoder auto generates the image embeddings from the CLIP adapter and passes the images as-is, which works. However, the `p_losses` pass then performs `normalize_neg_one_to_one` which shifts the range to `[-3, 1]`.

If I leave the images in the range `[0, 1]` then the `forward` pass in the Decoder passes the wrong image data into the CLIP adapter while the `p_losses` pass is correct with `[-1, 1]`.

As a workaround I can compute image_embed/text_embed manually and pass them into the Decoder, but this takes up extra memory or performance from having to discard the normalized results after and recompute them.

I think the control to disable auto-normalization would be beneficial here, but maybe this is more of an error training with normalized image data on CLIP itself.

**Comment by lucidrains** [comment]
ahh, i know how to fix this, but all locked up this week. i'll get to it in the event i do not accept work in the city

**Comment by lucidrains** [comment]
@marunine threw in a quick fix for you! https://github.com/lucidrains/DALLE2-pytorch/commit/db0642c4cdcaf4639136c1978c8124235d75a52e#diff-038ecede954c29266888cb88e37cc06f61ba2433f8b5142c7b2b2cefde5ed0edR1701 just set this to `False`

**Comment by lucidrains** [comment]
i realized another option is to allow one to pass in a `normalization_fn` into the `XClipAdapter`, but let's go with this for now

**Comment by marunine** [comment]
Thanks for the quick response!

I think the only issue with a `normalization_fn` with the XClipAdapter is that it implies that the image data would need to be normalized. I'm not sure if the model is more numerically stable or trains better if it's in `[-1, 1]` vs `[0, 1]`, but anecdotally it worked for me.

**Comment by lucidrains** [comment]
@marunine great! if you get anything working with x-clip as the base, do send me an email and let me know 😄

**Comment by jacobwjs** [comment]
> I have an x-clip model trained on normalized image data in the range `[-1, 1]` with good loss performance on a custom dataset.
> 

Currently making a run at a custom dataset using x-clip. Curious to know what params (learning rate, etc.) did you use?
